Code/GraphMol/SubstructLibrary/SubstructLibrary.h

```diff
@@ -0,0 +1,310 @@
+  MolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+    mols.push_back( boost::make_shared<ROMol>(m) );
+    return size()-1;
+  }
+    
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    return mols[idx];
+  }
+  
+  virtual unsigned int size() const { return rdcast<unsigned int>(mols.size()); }
+
+};
+
+//! Concrete class that holds binary cached molecules in memory
+/*!
+  This implementation uses quite a bit less memory than the
+  non cached implementation.  However, due to the reduced speed
+  it should be used in conjunction with a pattern fingerprinter.
+
+  See RDKit::FPHolder
+*/
+class CachedMolHolder: public MolHolderBase {
+  std::vector<std::string> mols;
+public:
+  CachedMolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+    mols.push_back( std::string() );
+    MolPickler::pickleMol(m, mols.back());
+    return size()-1;
+  }
+
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    boost::shared_ptr<ROMol> mol(new ROMol);
+    MolPickler::molFromPickle(mols[idx], mol.get());
+    return mol;
+  }
+  
+  virtual unsigned int size() const { return rdcast<unsigned int>(mols.size()); }
+};
+
+//! Concrete class that holds smiles strings in memory
+/*!
+    This implementation uses quite a bit less memory than the
+    cached binary or uncached implementation.  However, due to the
+    reduced speed it should be used in conjunction with a pattern
+    fingerprinter.
+
+    See RDKit::FPHolder
+*/
+class CachedSmilesMolHolder: public MolHolderBase {
+  std::vector<std::string> mols;
+public:
+  CachedSmilesMolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+
+    bool doIsomericSmiles = true;
+    mols.push_back(MolToSmiles(m, doIsomericSmiles));
+    return size()-1;
+  }
+  
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    boost::shared_ptr<ROMol> mol(SmilesToMol(mols[idx]));
+    return mol;
+  }
+  
+  virtual unsigned int size() const { return rdcast<unsigned int>(mols.size()); }
+
+};
+
+
+//! Base FPI for the fingerprinter used to rule out impossible matches
+class FPHolderBase {
+  std::vector<ExplicitBitVect *> fps;
+
+public:
+ virtual ~FPHolderBase() {
+   for(size_t i=0;i<fps.size();++i)
+     delete fps[i];
+   
+ }
+
+ //! Add a molecule to the fingerprinter
+ unsigned int addMol( const ROMol &m) {
+   fps.push_back(getQueryBits(m));
+   return rdcast<unsigned int>(fps.size() - 1);
+ }
+
+ //! Return false if a substructure search can never match the molecule
+ bool passesFilter(unsigned int idx, const ExplicitBitVect &query) const {
+   PRECONDITION(idx < fps.size(), "idx out of range in PatternHolder::passesFilter");
+   return AllProbeBitsMatch(query, *fps[idx]);
+ }
+ 
+ virtual ExplicitBitVect *getQueryBits(const ROMol &m) const = 0;
```

Comment:
Indeed it should.  I'll also rename this to me "makeQueryBits" to be more clear.